bindata/network/ovn-kubernetes-ipsec/ovn-ipsec.yaml

```diff
@@ -0,0 +1,207 @@
+                  -out ipsec-req.pem
+
+          csr_64=$(cat ipsec-req.pem | base64 | tr -d "\n")
+
+          # The signer controller does not allow re-signing a key. We will
+          # delete the old key to be sure it is not there
+          kubectl delete --ignore-not-found=true csr/$(hostname)
+
+          cat <<EOF | kubectl apply -f -
+          apiVersion: certificates.k8s.io/v1
+          kind: CertificateSigningRequest
+          metadata:
+            name: $(hostname)
+          spec:
+            request: ${csr_64}
+            signerName: network.openshift.io/signer
+            usages:
+            - ipsec tunnel
+          EOF
+
+          until [ ! -z $(kubectl get csr/$(hostname) -o jsonpath='{.status.certificate}' 2>/dev/null) ]
+          do
+            ((counter++))
+            if [ $counter -gt 60 ];
+            then
+                    echo "Unable to sign certificate after $counter seconds"
+                    exit 1
+            fi
+          done
+
+          kubectl get csr/$(hostname) -o jsonpath='{.status.certificate}' | base64 -d | openssl x509 -outform pem -text -out ipsec-cert.pem
+
+          kubectl delete csr/$(hostname)
+
+          cat /signer-ca/tls.crt | openssl x509 -outform pem -text > /etc/openvswitch/keys/ipsec-cacert.pem
+
+          ovs-vsctl --retry -t 60 set Open_vSwitch . other_config:certificate=/etc/openvswitch/keys/ipsec-cert.pem \
+                                                     other_config:private_key=/etc/openvswitch/keys/ipsec-privkey.pem \
+                                                     other_config:ca_cert=/etc/openvswitch/keys/ipsec-cacert.pem
+
+          popd
+
+          # Workaround for https://github.com/libreswan/libreswan/issues/373
+          ulimit -n 1024
+
+          /usr/libexec/ipsec/addconn --config /etc/ipsec.conf --checkconfig
+          # Check for kernel modules
+          /usr/libexec/ipsec/_stackmanager start
+          # Check for nss database status and migration
+          /usr/sbin/ipsec --checknss
+          # Check for nflog setup
+          /usr/sbin/ipsec --checknflog
+          # Start the actual IKE daemon
+          /usr/libexec/ipsec/pluto --leak-detective --config /etc/ipsec.conf --logfile /var/log/openvswitch/libreswan.log
+
+          # Workaround for https://mail.openvswitch.org/pipermail/ovs-dev/2020-October/375734.html
+          OVS_LOGDIR=/var/log/openvswitch OVS_RUNDIR=/var/run/openvswitch OVS_PKGDATADIR=/usr/share/openvswitch /usr/share/openvswitch/scripts/ovs-ctl --ike-daemon=libreswan start-ovs-ipsec
+
+          # Wait for the nbdb container in the ovnkube-master pod to become ready before we try to write to
+          # it.
+          if kubectl wait --for=condition=Ready pod -l app=ovnkube-master -n ${ovn_kubernetes_namespace} --timeout 600s; then
+            echo "ovnkube-master is not Ready, unable to write to OVNNB DB"
+            exit 2
+          else
+            ovn-nbctl set nb_global . ipsec=true
```

Comment:
Don't touch the NBDB in the ovnkube-node database - this needs to be somewhere else.

Reply:
I wanted to do this in the nbdb in ovnkube-master, however, due to the way that pod is started (using ovn-ctl run_nb_ovsdb - which blocks) i couldn't easily template it in. Is this the correct place or could you suggest elsewhere?

Reply:
This isn't the right place; it needs to be in the master pod somewhere.

If it makes sense, it can be a PostStartHook on the correct container. Or it can be its own container. that configures ipsec and sleeps forever.

Reply:
Yeah, I was thinking about "PostStartHook" yesterday evening and I have implemented that successfully.